otri/validation/__init__.py

```diff
@@ -0,0 +1,572 @@
+        `append_label` method of this module.
+        '''
+        append_label(data, label)
+
+
+class LinearValidator(ValidatorFilter):
+
+    '''
+    This class extends `ValidatorFilter` and is meant to be used when running checks on single atoms
+    coming from Streams. It simply checks the atoms one at a time and then outputs them on the
+    output stream with the same index as the input.
+
+    Must have the same number of inputs and outputs.
+
+    This class can be fully implemented by overriding `_check()` alone.
+    '''
+
+    def __init__(self, inputs: Sequence[str], outputs: Sequence[str], check: Callable = None):
+        '''
+        This Validator expects the same number of Stream inputs and outputs.
+
+        Parameters:
+            inputs : str
+                Names of the inputs.
+
+            outputs : str
+                Names of the outputs.
+
+            check : Callable
+                If you don't want to override the class, you can pass a `Callable` here.
+                The `Callable` should expect a single atom as a parameter.
+        '''
+        # Check same amount of inputs and outputs.
+        if len(inputs) != len(outputs):
+            raise ValueError("The number of input and output Streams must be the same.")
+
+        super().__init__(inputs, outputs, len(inputs), len(outputs))
+
+        # Overwrite check.
+        if check:
+            self._check = check
+
+    def _on_ok(self, data: Mapping, index: int):
+        '''
+        Called if data resulted ok.
+        Pushes the atom to the output on the same index it came from.
+
+        Parameters:
+            data : Mapping
+                The checked data.
+        '''
+        self._push_data(data, index)
+
+    def _on_result(self, data: Mapping, result: Any, index: int):
+        '''
+        Called if `_check` returned something. Labels the atom and pushes it to the output.
+
+        Parameters:
+            data : Mapping
+                The checked data.
+
+            result : Any
+                The result of `_check`.
+
+            index : int
+                The index of the input the data has been popped from.
+        '''
+        self._add_label(data, result)
+        self._push_data(data, index)
+
+
+class MonoValidator(LinearValidator):
+    '''
+    A `LinearValidator` for a single Stream.
+    '''
+
+    def __init__(self, inputs: str, outputs: str, check: Callable = None):
+        '''
+        Parameters:
+            inputs : str
+                Name for the single input stream.
+
+            outputs : str
+                Name for the single output stream.
+
+            check : Callable
+                If you don't want to override the class, you can pass a Callable here.
+                The Callable should require only the atom as a parameter.
+        '''
+        super().__init__([inputs], [outputs], check)
+
+
+class BufferedValidator(LinearValidator):
```

Comment:
Codacy found an issue: [Method '_check' is abstract in class 'ValidatorFilter' but is not overridden](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=6277039)